lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -283,58 +283,75 @@ static bool isBound(hw::HWModuleLike op, hw::InstanceGraph &instanceGraph) {
-  if (isBound(oldMod, instanceGraph)) {
-    LLVM_DEBUG(oldMod->emitWarning("already bound, skipping for ") << bindFile);
-    return;
-  }
-
-  // Move the module to the test code path.
-  oldMod->setAttr("output_file", testBenchDir);
-
-  // Replace its instances with binds.
+  // Iterate through each of the module's inputs.
   hw::InstanceGraphNode *node = instanceGraph.lookup(oldMod);
-  OpBuilder b = OpBuilder(oldMod.getContext());
+  OpBuilder b(oldMod);
+  bool allInlined = true;
   for (hw::InstanceRecord *use : node->uses()) {
-    OpBuilder::InsertionGuard g(b);
-
+    // If there is no instance, move on.
     hw::HWInstanceLike instLike = use->getInstance();
-    if (!instLike)
+    if (!instLike) {
+      allInlined = false;
       continue;
+    }
 
-    // Mark the instance to not be printed.
+    // If the instance had a symbol, we can't inline it.
     hw::InstanceOp inst = cast<hw::InstanceOp>(instLike.getOperation());
-    hw::HWModuleLike parent = use->getParent()->getModule();
-    inst->setAttr("doNotPrint", b.getBoolAttr(true));
-
-    // Give the instance an inner sym if it needs one.
-    if (!inst.getInnerSym().has_value())
-      inst.setInnerSymAttr(
-          b.getStringAttr("__" + inst.getInstanceName() + "__"));
-
-    // Create the bind.
-    b.setInsertionPointToEnd(
-        &oldMod->getParentOfType<mlir::ModuleOp>()->getRegion(0).front());
-    auto bindOp = b.create<sv::BindOp>(inst.getLoc(), parent.moduleNameAttr(),
-                                       inst.getInnerSymAttr());
-    if (bindFile)
-      bindOp->setAttr("output_file", bindFile);
+    if (inst.getInnerSym().has_value()) {
+      allInlined = false;
+      continue;
+    }
+
+    // Build a mapping from module block arguments to instance inputs.
+    BlockAndValueMapping mapping;
+    assert(inst.getInputs().size() == oldMod.getNumInputs());
+    auto inputPorts = oldMod.getBodyBlock()->getArguments();
+    for (size_t i = 0, e = inputPorts.size(); i < e; ++i)
+      mapping.map(inputPorts[i], inst.getOperand(i));
+
+    // Inline the body at the instantiation site.
+    hw::HWModuleOp instParent =
+        cast<hw::HWModuleOp>(use->getParent()->getModule());
+    OpBuilder::InsertionGuard g(b);
+    b.setInsertionPoint(inst);
+    for (auto &op : llvm::make_early_inc_range(*oldMod.getBodyBlock())) {
+      // For instances in the bind table, update the bind with the new parent.
+      if (auto innerInst = dyn_cast<hw::InstanceOp>(op)) {
+        if (auto innerInstSym = innerInst.getInnerSymAttr()) {
+          auto it = bindTable[oldMod.getNameAttr()].find(innerInstSym);
+          if (it != bindTable[oldMod.getNameAttr()].end()) {
+            sv::BindOp bind = it->second;
+            auto oldInnerRef = bind.getInstanceAttr();
+            auto newInnerRef = hw::InnerRefAttr::get(
+                instParent.moduleNameAttr(), oldInnerRef.getName());
+            bind.setInstanceAttr(newInnerRef);
+          }
+        }
+      }
```

Comment:
When inlining modules, do we have to re-unique any inner-symbols, and fixup any InnerRef's which point to them?

Reply:
I'm not sure... I fixed up the binds, but I guess there could be something else? I forget the status in the HW dialect, hopefully we have the attribute to point to any hw.globalRefs that point to the instance.

Reply:
Oh actually, I explicitly block inlining if there is a symbol on the instance: https://github.com/llvm/circt/pull/4063/files#diff-420e195caff1364a2e07efc8d4f14978ee98a7f3b5533583ee1f10f9f9f88561R308. My thought process was if anything referred to that non-locally, better to not inline it. In practice, instances of the modules we want inlined here shouldn't have that symbol.